src/hotspot/share/gc/shenandoah/heuristics/shenandoahReactiveHeuristics.cpp

```diff
@@ -0,0 +1,308 @@
+#include "utilities/quickSort.hpp"
+
+// These constants are used to adjust the margin of error for the moving
+// average of the allocation rate and cycle time. The units are standard
+// deviations.
+const double ShenandoahReactiveHeuristics::FULL_PENALTY_SD = 0.2;
+const double ShenandoahReactiveHeuristics::DEGENERATE_PENALTY_SD = 0.1;
+
+// These are used to decide if we want to make any adjustments at all
+// at the end of a successful concurrent cycle.
+const double LOWEST_EXPECTED_AVAILABLE_AT_END = -0.5;
+const double HIGHEST_EXPECTED_AVAILABLE_AT_END = 0.5;
+
+// These values are the confidence interval expressed as standard deviations.
+// At the minimum confidence level, there is a 25% chance that the true value of
+// the estimate (average cycle time or allocation rate) is not more than
+// MINIMUM_CONFIDENCE standard deviations away from our estimate. Similarly, the
+// MAXIMUM_CONFIDENCE interval here means there is a one in a thousand chance
+// that the true value of our estimate is outside the interval. These are used
+// as bounds on the adjustments applied at the outcome of a GC cycle.
+const double ShenandoahReactiveHeuristics::MINIMUM_CONFIDENCE = 0.319; // 25%
+const double ShenandoahReactiveHeuristics::MAXIMUM_CONFIDENCE = 3.291; // 99.9%
+
+ShenandoahReactiveHeuristics::ShenandoahReactiveHeuristics() :
+  ShenandoahAdaptiveHeuristics(),
+  _allocation_rate(this),
+  _available_at_cycle_start(0),
+  _margin_of_error_sd(ShenandoahReactiveInitialConfidence),
+  _spike_threshold_sd(ShenandoahReactiveInitialSpikeThreshold),
+  _last_trigger(OTHER) { }
+
+ShenandoahReactiveHeuristics::~ShenandoahReactiveHeuristics() {}
+
+void ShenandoahReactiveHeuristics::record_cycle_start() {
+  ShenandoahAdaptiveHeuristics::record_cycle_start();
+  _allocation_rate.allocation_counter_reset();
+  _available_at_cycle_start = ShenandoahHeap::heap()->free_set()->available();
+}
+
+bool ShenandoahReactiveHeuristics::should_start_gc() const {
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  size_t capacity = heap->max_capacity();
+  size_t available = heap->free_set()->available();
+  size_t bytes_allocated_since_gc_start = heap->bytes_allocated_since_gc_start();
+
+  // Track allocation rate even if we decide to start a cycle for other reasons.
+  ShenandoahReactiveHeuristics *heuristic = const_cast<ShenandoahReactiveHeuristics *>(this);
+  heuristic->_allocation_rate.sample(bytes_allocated_since_gc_start);
+  heuristic->_last_trigger = OTHER;
+
+  if (is_available_below_min_threshold(capacity, available)) {
+    return true;
+  }
+
+  if (is_learning_necessary(capacity, available)) {
+    return true;
+  }
+
+  if (is_allocation_rate_too_high(capacity, available, bytes_allocated_since_gc_start)) {
+    return true;
+  }
+
+  // Don't call into our immediate parent class, we've already done
+  // everything it would do (and more).
+  return ShenandoahHeuristics::should_start_gc(); // NOLINT(bugprone-parent-virtual-call)
```

Comment:
We don't do lint (disable) warnings in the source code. I understand it might be useful for bugprone, but we don't do that anyway. This should go away once changes are merged with "adaptive", AFAICS.

Reply:
Yes, this is gone now.